src/HexagonOffload.cpp

```diff
@@ -0,0 +1,322 @@
+        Expr& var = state_vars[name];
+        if (!var.defined()) {
+            Buffer storage(type, {}, nullptr, name + "_buf");
+            *(void **)storage.host_ptr() = nullptr;
+            var = Load::make(type_of<void*>(), name + "_buf", 0, storage, Parameter());
+        }
+        return var;
+    }
+
+    Expr state_var_ptr(const std::string& name, Type type) {
+        Expr var = state_var(name, type);
+        return Call::make(Handle(), Call::address_of, {var}, Call::Intrinsic);
+    }
+
+    Expr module_state() {
+        return state_var("hexagon_module_state", type_of<void*>());
+    }
+
+    Expr module_state_ptr() {
+        return state_var_ptr("hexagon_module_state", type_of<void*>());
+    }
+
+    // Create a Buffer containing the given buffer/size, and return an
+    // expression for a pointer to the first element.
+    Expr buffer_ptr(const uint8_t* buffer, size_t size, const char* name) {
+        Buffer code(type_of<uint8_t>(), {(int)size}, nullptr, name);
+        memcpy(code.host_ptr(), buffer, (int)size);
+
+        Expr ptr_0 = Load::make(type_of<uint8_t>(), name, 0, code, Parameter());
+        return Call::make(Handle(), Call::address_of, {ptr_0}, Call::Intrinsic);
+    }
+
+    Stmt call_extern_and_assert(const std::string& name, const std::vector<Expr>& args) {
+        Expr call = Call::make(Int(32), name, args, Call::Extern);
+        string call_result_name = unique_name(name + "_result");
+        Expr call_result_var = Variable::make(Int(32), call_result_name);
+        return LetStmt::make(call_result_name, call,
+                             AssertStmt::make(EQ::make(call_result_var, 0), call_result_var));
+    }
+
+    using IRMutator::visit;
+
+    void visit(const For *loop) {
+        if (loop->device_api == DeviceAPI::Hexagon) {
+            // Unrolling or loop partitioning might generate multiple
+            // loops with the same name, so we need to unique them.
+            std::string hex_name = "hex_" + loop->name;
+
+            // After moving this to Hexagon, it doesn't need to be
+            // marked Hexagon anymore.
+            Stmt body = For::make(loop->name, loop->min, loop->extent, loop->for_type,
+                                  DeviceAPI::None, loop->body);
+            body = remove_trivial_for_loops(body);
+
+            // Build a closure for the device code.
+            // TODO: Should this move the body of the loop to Hexagon,
+            // or the loop itself? Currently, this moves the loop itself.
+            Closure c(body);
+
+            // Make an argument list, and generate a function in the
+            // device_code module. The hexagon runtime code expects
+            // the arguments to appear in the order of (input buffers,
+            // output buffers, input scalars).  There's a huge hack
+            // here, in that the scalars must be last for the scalar
+            // arguments to shadow the symbols of the buffer.
```

Comment:
(1) Does this huge hack warrant a TODO to fix later, or is it a permanent hack?
(2) Is there reality checking (here or elsewhere) that verifies that the scalars are in fact last?